cmd/jwt_parser.go

```diff
@@ -0,0 +1,173 @@
+/*
+ * MinIO Cloud Storage, (C) 2020 MinIO, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package cmd
+
+// This file is a re-implementation of the original code here with some
+// additional allocation tweaks reproduced using GODEBUG=allocfreetrace=1
+// original file https://github.com/dgrijalva/jwt-go/blob/master/parser.go
+// borrowed under MIT License https://github.com/dgrijalva/jwt-go/blob/master/LICENSE
+
+import (
+	"crypto"
+	"crypto/hmac"
+	"encoding/base64"
+	"fmt"
+	"strings"
+	"sync"
+
+	jwtgo "github.com/dgrijalva/jwt-go"
+	jsoniter "github.com/json-iterator/go"
+)
+
+// SigningMethodHMAC - Implements the HMAC-SHA family of signing methods signing methods
+// Expects key type of []byte for both signing and validation
+type SigningMethodHMAC struct {
+	Name string
+	Hash crypto.Hash
+}
+
+// Specific instances for HS256 and company
+var (
+	SigningMethodHS256 *SigningMethodHMAC
+	SigningMethodHS384 *SigningMethodHMAC
+	SigningMethodHS512 *SigningMethodHMAC
+)
+
+var hmacSigners = []*SigningMethodHMAC{
+	{"HS256", crypto.SHA256},
+	{"HS384", crypto.SHA384},
+	{"HS512", crypto.SHA512},
+}
+
+type keyFunc func(*MapClaims) ([]byte, error)
+
+// ParseWithClaims - parse the token string, valid methods.
+func ParseWithClaims(tokenStr string, claims *MapClaims, fn keyFunc) error {
+	bufp := spool.Get().(*[]byte)
+	defer spool.Put(bufp)
+
+	signer, err := ParseUnverified(tokenStr, claims, *bufp)
+	if err != nil {
+		return err
+	}
+
+	// Lookup key
+	if fn == nil {
+		// keyFunc was not provided.  short circuiting validation
+		return jwtgo.NewValidationError("no Keyfunc was provided.", jwtgo.ValidationErrorUnverifiable)
+	}
+
+	key, err := fn(claims)
+	if err != nil {
+		return err
+	}
+
+	// Validate Claims
+	if err = claims.Valid(); err != nil {
+		// If the Claims Valid returned an error, check
+		// if it is a validation error, If it was another
+		// error type, create a jwtgo.ValidationError with
+		// a generic ClaimsInvalid flag set
+		return err
+	}
```

Comment:
I am bit skeptical of returning an invalid claims error before the hash is verified - I dont think this is cryptographically sound. The JWT standard also says the same - claims should not be touched unless the signature is found to be sound. Should we do a kind of security audit of this code? Perhaps you don't have to do it like this to get the same performance gains.

Reply:
That is what original jwt-go library does @donatello

Reply:
The performance gains are not associated with validation here those are associated with optimal buffer usage, so validation piece has not changed from the original author. If we have been using jwt-go then we should be fine @donatello

Reply:
This is not a good idea to copy over so much crypto code that we would not be actively maintaining or finding security problems with.

For example if you look at the original author's new implementation (Release 4.x), the order of claim validation and signature verification is reversed as mentioned in the JWT standard - https://github.com/dgrijalva/jwt-go/blob/release_4_0_0/parser.go#L74 - this is an important point because if the signature is invalid and the claims are invalid, we return a claims are invalid error and that may lead to implementation error somehow implying that the signature is in-fact valid.

Reply:
This is a tangential discussion @donatello I don't think this PR mentioned that it is trying to fix the security constraints in the original library - if 4.x addresses this we shall move when it's released and see if has any gains in performance. 

So the PR is not improving anything related to security, just that it's tweaking the original implementation with some performance gains. I can do the change that is done in 4.x right now but it is a tangential discussion - the code is simply borrowed as is and will not improve on the security front for us. So let me know it is a quick change anyways on our end.

Reply:
No, that's right that the goal of this PR is not to fix the crypto implementation. But pulling in the internal code of the library makes it so that even when the new release of the the jwt library happens we may not fix this code. Maybe we can leave a FIXME note in the code and move on if this tradeoff needs to be made in favor of performance.

Reply:
I just moved the code around which seems to work anyways